New Invoice Type: Splitter Escrow (Smart Contract and Testing)

geovgy opened the pull request:
This PR includes the new invoice type to be used by Raid Guild for client work based on the scope from the [Smart Escrow RIP](https://github.com/raid-guild/RIPs/issues/130). A few things that should be noted:
- The encoded `_data` param includes two new variables that are decoded in the `_handleData` function. The first is `dao` which is the address of the dao treasury that will receive any spoils. The amount of spoils is determined by the second new variable, `daoFee`. This is the same as split factor in the wrapped smart invoice contract. It is a `uint256` variable and the basis points (out of 10000) for the % of spoils the DAO collects.
- On release, the tokens are automatically split based on the `daoFee` and transfers the split amounts to the `provider` and `dao` addresses. This removes the additional step required by wrapped smart invoice to disperse the splits.
- Milestone releases no longer need to be in sequential order. If a client wants to release a specific milestone without releasing others indexed earlier, they can now. `milestoneReleased` is a mapping that indicates whether a milestone has been released or not, so then clients can release in any arbitrary order. This is helpful if milestones are based on particular roles (like dev, design, etc.) rather than timeline.
          - `release()` determines milestone to be released by first index marked `false` in `milestoneReleased`.
          - `release(uint256 _milestone)` releases only releases that milestone and no others.
          - `release(uint256[] _milestones)` can release multiple milestones at once. This functions similar to how `release(uint256 _milestone)` used to but with more control of which milestones.

I will comb through the contract for clean up and audit before merging. The ABI of the main functions, however, should be ready to use for frontend integration.

plor commented on packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:
```diff
@@ -0,0 +1,587 @@
+     * Uses the internal `_release` function to perform the actual release.
+     * @param _milestones The milestones to release funds
+     */
+    function release(uint256[] calldata _milestones) external nonReentrant {
```
This looks quite complicated to me. Would it be possible to separate the PR with the splitting and the release bit? I prefer when PRs tackle one feature and I'm not sure this is necessary for splitting. I would personally hold off for further refinement on the milestone release bits, but I think the splitting is important to get merged in sooner.

plor commented on packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:
```diff
@@ -0,0 +1,587 @@
+        }
+        require(balance >= amount, "insufficient balance");
+
+        uint256 daoAmount = (amount * daoFee) / 10000;
+        uint256 providerAmount = amount - daoAmount;
+        IERC20(token).safeTransfer(dao, daoAmount);
+        IERC20(token).safeTransfer(provider, providerAmount);
+        released = released + amount;
+        _remaining = remaining;
```
This is repeated logic. Couldn't this be shared in some way?

plor reviewed (changes requested):
I'll leave it to @psparacino to give his views, but I feel like this will give an inconsistent ux between different invoice types that is not appealing and tricky to manage on the UI side.

spengrah commented on packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:
```diff
@@ -0,0 +1,587 @@
+        }
+
+        uint256 balance = IERC20(token).balanceOf(address(this));
+        uint256 amount = 0;
```
FYI its not necessary to instantiate vars explicitly at 0. All vars are automatically initiated at the null value for their type, eg `uint256 amount;` is the same thing as `uint356 amount = 0;`

spengrah replied:
also, you're looping through the same array twice. If you do the two checks on lines 354 and 355 as part of the loop starting on 365, you can get rid of that first loop.

spengrah commented on packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:
```diff
@@ -0,0 +1,587 @@
+        if (resolverType == ADR.ARBITRATOR) {
+            disputeId = IArbitrator(resolver).createDispute{value: msg.value}(
+                NUM_RULING_OPTIONS,
+                abi.encodePacked(details)
```
I know this is probably not in the scope of the added functionality, but should this be the `_details` function arg instead of the `details` state var?

spengrah commented on packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:
```diff
@@ -0,0 +1,587 @@
+
+        uint256 balance = IERC20(token).balanceOf(address(this));
+
+        if (_remaining >= 1) {
```
FYI performing the else conditional (`_remaining < 1`) first will save a bit of gas by using 1 less operation.

spengrah commented on packages/contracts/contracts/SmartInvoiceSplitEscrow.sol:
```diff
@@ -0,0 +1,587 @@
+            // milestone = milestone + 1;
+            milestoneReleased[_milestone] = true;
+            uint256 daoAmount = (amount * daoFee) / 10000;
+            uint256 providerAmount = amount - daoAmount;
```
this can be unchecked since daoAmount must always be < amount

geovgy commented:
Alright, I updated the contract based on some of your suggestions (both of you). The major ones are the following:
- I removed the arbitrary milestones and have it in a separate branch.
- It now inherits `SmartInvoiceEscrow` directly. You'll notice updates on `SmartInvoiceEscrow` too which are just adding `virtual` to some functions in order to override them in `SmartInvoiceSplitEscrow`.
- Refactored the split transfers into its own internal function

A minor change I also made was reordering the require statements in `_handleData` function. Since the `_resolutionRate` is always set to a non-zero value regardless, seemed unnecessary to require a check for it.

geovgy commented:
I did some more refactoring so now the `SmartInvoiceSplitEscrow` only overrides two functions, `_handleData` and `_transferPayment`. Much simpler and cleaner from before. Credit to @plor for the suggestion.

plor commented:
Just checking that you were able to address all of @spengrah 's comments?

plor reviewed (approved):
These changes address my concerns.

geovgy commented:
> Just checking that you were able to address all of @spengrah 's comments?

All except one about using the _details arg vs details state variable for one of the arbitration functions: https://github.com/SmartInvoiceXYZ/smart-invoice/pull/125#discussion_r1225958880.
